src/Umbraco.Web/Extensions/UdiExtensions.cs

```diff
@@ -16,24 +17,20 @@ public static IPublishedContent ToPublishedContent(this Udi udi)
             var guidUdi = udi as GuidUdi;
             if (guidUdi == null) return null;
 
+            var umbracoContext = UmbracoContext.Current;
+            if (umbracoContext == null) return null;
+
             var umbracoType = Constants.UdiEntityType.ToUmbracoObjectType(guidUdi.EntityType);
 
-            var umbracoHelper = new UmbracoHelper(UmbracoContext.Current);
-            var entityService = ApplicationContext.Current.Services.EntityService;
             switch (umbracoType)
             {
                 case UmbracoObjectTypes.Document:
-                    return umbracoHelper.TypedContent(guidUdi.Guid);
+                    return umbracoContext.ContentCache.GetById(guidUdi.Guid);
                 case UmbracoObjectTypes.Media:
-                    var mediaAttempt = entityService.GetIdForKey(guidUdi.Guid, umbracoType);
-                    if (mediaAttempt.Success)
-                        return umbracoHelper.TypedMedia(mediaAttempt.Result);
-                    break;
+                    return umbracoContext.MediaCache.GetById(guidUdi.Guid);
                 case UmbracoObjectTypes.Member:
-                    var memberAttempt = entityService.GetIdForKey(guidUdi.Guid, umbracoType);
```

Comment:
The `MembershipHelper` allows direct querying on key, so no need to get the id first!